### PR TITLE
support dynamic resolution image encoding for Nemotron Nano VL

### PR DESCRIPTION
**Pull request by netanel-haber**
Adds dynamic image resolution support for encoding images with Nemotron Nano VL, while preserving static resolution images as-is. This means that images contribute a variable number of tokens to the prompt.  
Dynamic resolution applies to images only; **video remains static resolution**.

Adapted from code and work of, among others:
https://github.com/tylerpoon
https://github.com/trintamaki

Dynamic resolution can be enabled by adding `min_num_patches` and `max_num_patches` to config["vision_config"]["args"].

Aside from adding the dynamic tiler in `vllm/model_executor/models/nano_nemotron_vl.py` processing code, the PR expands `radio.py` and `intern_vit.py` to support a concatenated ragged tensor of images, using manual attention masking so that images don't attend to each other.

### CODE REVIEW

**Review by tlrmchlsmth — COMMENTED** [review]
>Adapted from code and work of, among others:
https://github.com/tylerpoon
https://github.com/trintamaki

[netanel-haber](https://github.com/netanel-haber) could you links to the code that was adapted? E.g. "this code was adapted from ..." 

Need to check licensing as well

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 1]
```diff
@@ -267,6 +290,327 @@ def calculate_timestamps(
...
+            # else try with the scaled down num_tokens_available_per_media.
+            if not scaled_down:
+                num_tokens_available_per_media = [self._min_num_patches] * len(
+                    media_list
+                )
+            else:
+                num_tokens_available_per_media = (
+                    scaled_down_num_tokens_available_per_media
+                )
+        assert_never(num_tokens_available_per_media)
```
Comment (tlrmchlsmth) [inline vllm/model_executor/models/nano_nemotron_vl.py]: This looks wrong, please replace with a proper assert or raise an exception

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 2]
```diff
@@ -1239,36 +1629,49 @@ def extract_feature(self, pixel_values):
...
+            num_tokens_per_image = kwargs["num_tokens_per_image"]
+            imgs_sizes = [(pv.shape[1], pv.shape[2]) for pv in pixel_values_flat]
+            pixel_values_flat = DynamicResolutionImageTiler.stack(
+                pixel_values_flat, self.patch_size
+            )
+            return NanoNemotronVLImagePixelInputsDynamic(
+                pixel_values_flat=pixel_values_flat,
+                imgs_sizes=imgs_sizes,
+                num_tokens_per_image=num_tokens_per_image,
+                validate=False,  # TODO(nhaber)
```
Comment (tlrmchlsmth) [inline vllm/model_executor/models/nano_nemotron_vl.py]: What is the TODO? Could you change the comment to be more specific? Or otherwise address in this PR

**Comment by tlrmchlsmth** [comment]
@ywang96 or @DarkLight1337 could you take a look?

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 3]
```diff
@@ -292,15 +637,32 @@ def __init__(
...
+            (image_size // patch_size) ** 2 * (downsample_ratio**2)
         )
         self.image_size = image_size
         self.use_thumbnail: bool = config.use_thumbnail
         self.norm_mean = torch.Tensor(config.norm_mean).reshape(1, 3, 1, 1)
         self.norm_std = torch.Tensor(config.norm_std).reshape(1, 3, 1, 1)
 
+        self.dynamic_tiler: DynamicResolutionImageTiler | None = None
+        if self.use_dynamic_resolution(config):
+            self.dynamic_tiler: DynamicResolutionImageTiler = (
```
Comment (DarkLight1337) [inline vllm/model_executor/models/nano_nemotron_vl.py]:
```suggestion
            self.dynamic_tiler = (
```

Unnecessary type annotation

File: vllm/model_executor/models/intern_vit.py [hunk 4]
```diff
@@ -223,15 +223,28 @@ def _apply_qk_norm(self, q: torch.Tensor, k: torch.Tensor):
...
+            q = q.view(B, N, self.num_heads_per_partition, self.head_dim)
+            k = k.view(B, N, self.num_heads_per_partition, self.head_dim)
+            v = v.view(B, N, self.num_heads_per_partition, self.head_dim)
+            q, k, v = (t.transpose(1, 2) for t in (q, k, v))
+            out = F.scaled_dot_product_attention(
+                q, k, v, attn_mask=attn_mask, scale=self.scale
+            )
+            out = out.transpose(1, 2).reshape(B, N, -1)
+        else:
+            out = self.attn(q, k, v)
```
Comment (Isotr0py) [inline vllm/model_executor/models/intern_vit.py]: I prefer to decouple it through `RadioVisionEncoderLayer` instead of updating InternVL's ViT directly, especially this `attn_mask` is not used by InternVL exactly.

File: vllm/model_executor/models/nano_nemotron_vl.py (hunk 2 above)
Comment (netanel-haber) [reply vllm/model_executor/models/nano_nemotron_vl.py]: Done

File: vllm/model_executor/models/intern_vit.py (hunk 4 above)
Comment (netanel-haber) [reply vllm/model_executor/models/intern_vit.py]: Done

File: vllm/model_executor/models/intern_vit.py [hunk 5]
```diff
@@ -274,6 +274,8 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
 
 
 class InternVisionEncoderLayer(nn.Module):
+    attn_cls: InternParallelAttention = InternParallelAttention
```
Comment (DarkLight1337) [inline vllm/model_executor/models/intern_vit.py]: This should be a parameter to `__init__`

Comment (netanel-haber) [reply vllm/model_executor/models/intern_vit.py]: Done

File: vllm/model_executor/models/radio.py [hunk 6]
```diff
@@ -406,6 +477,66 @@ def __init__(self, patch_size: int, embed_dim: int, bias: bool = False, **factor
...
+            hidden_states
+            + self.attn(self.norm1(hidden_states), attn_mask=attn_mask) * self.ls1
+        )
+
+        hidden_states = hidden_states + self.mlp(self.norm2(hidden_states)) * self.ls2
+
+        return hidden_states
+
+
+class RadioVisionEncoder(InternVisionEncoder):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/radio.py:525]: I think you don't need this class anymore, just initialize `InternVisionEncoder` directly with `layer_cls` argument

Comment (netanel-haber) [reply vllm/model_executor/models/radio.py:525]: I think not, since the encoder forward needs to be overriden as well to accept and process `attn_mask`, no?

Comment (DarkLight1337) [reply vllm/model_executor/models/radio.py:525]: Sorry I missed that, then it's fine.

**Review by DarkLight1337 — APPROVED** [review]

**Review by Isotr0py — APPROVED** [review]
